### ISSUE TITLE
cmd/compile: internal compiler error: assertion failed

### ISSUE THREAD

**Issue by r-hang** [golang/go#54135]
### What version of Go are you using (`go version`)?

master
$ go version
go version devel go1.19-5c8ec89cb5 Thu Jul 28 21:13:04 2022 +0000 linux/amd64

### Does this issue reproduce with the latest release?
Fails on both go1.19rc2 and master. Passes on go1.18.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="off"
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/user/.cache/go-build"
GOENV="/home/user/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/user/go-code2/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/user/go-code2"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org/,direct"
GOROOT="/home/user/go/src/github.com/golang/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/user/go/src/github.com/golang/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="devel go1.19-5c8ec89cb5 Thu Jul 28 21:13:04 2022 +0000"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3108614045=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

Try compiling this test program

```go
package main

import (
	"fmt"
	"testing"
)

type Bar[T any] interface{
    Blanker()
}

type FooBar[T any] interface{}

type Foo[E comparable] struct {
	me E
}

func (b *Foo[E]) do(h Bar[E], f FooBar[E]) {
	check(h, f)
}

func check[T comparable](p Bar[T], q FooBar[T]) int {
	if _, ok := p.(Checkable); ok {
		fmt.Println(ok)
	}
	if _, ok := q.(Checkable); ok {
		fmt.Println(ok)
	}
	return 1
}

type Checkable interface{}

func TestFunction(m *testing.T) {
	x := Foo[int]{me: 2}
	fmt.Println(x)
}
```

### What did you expect to see?

Test passed.

### What did you see instead?
./main_test.go:19:7: internal compiler error: assertion failed

**Comment by sywhang** [comment]
To add a bit of context, we found this issue while testing Go 1.19 on our Go Monorepo at Uber. We noticed both master and 1.19 rc2 is breaking with some of the latest changes in our codebase that made use of generics.

The program above is a small repro we were able to isolate out of the actual code that broke.

**Comment by ianlancetaylor** [comment]
May be a dup of #53762.

**Comment by sywhang** [comment]
@ianlancetaylor thanks, i can confirm the failure is coming from the same place as #53762. 

The repro program we posted above should help debugging the root cause.

**Comment by ianlancetaylor** [comment]
Thanks, closing as a dup.  My understanding is that #53762 has already been fixed on tip.

**Comment by sywhang** [comment]
@ianlancetaylor Sorry, but I don't think it has been. The program above still fails to build on tip. The latest comments in #53762 seems like there are others facing the same issue on tip. 

[> Will you re-open this issue or should I create a new one?](https://github.com/golang/go/issues/53762#issuecomment-1183581889)

**Comment by ianlancetaylor** [comment]
You're right, it's still crashing.  Sorry about that.

CC @randall77 @mdempsky

**Comment by sywhang** [comment]
Tested the fix above, and it still is failing to build some generics code. I will follow up with a repro.

**Comment by sywhang** [comment]
The repro below still fails to build with the fix above. 

```go
import (
        "fmt"
        "testing"
)

type Foo[T comparable] interface {
        FooFoo(tt []T)
}

type Bar interface {
        BarSize() int
}

func Test[T comparable](tt []T, foos []Foo[T]) {
        if len(foos) == 0 {
                return
        }
        p := FooBar[T]{fooCnt: len(foos)}
        p.getT(foos[0])
}

type FooBar[T comparable] struct {
        fooCnt int
}

func (p *FooBar[T]) getT(foo Foo[T]) {
        barSize(foo)
}

func barSize[T comparable](p Foo[T]) {
        s := 1
        if b, ok := p.(Bar); ok {
                s = b.BarSize()
        }
        fmt.Println(s)
}

func TestTest(t *testing.T) {
        tt := []int{1, 2, 3}
        foos := []Foo[int]{}

        Test(tt, foos)
}
```

On Go 1.18, this passes as expected.

On Go 1.19, both on tip and tip with the patch in https://go-review.googlesource.com/c/go/+/420394/

**Comment by sywhang** [comment]
Can this fix be cherry-picked to Go 1.19 branch before the release tomorrow? Plenty of code out there will be blocked on upgrading to 1.19 until this fix is made available.
